toolbox/relnotes/main.go

```diff
@@ -0,0 +1,757 @@
+// markdown file.
+func createBody(c *github.Client, f *os.File, releaseTag, branch, docURL, exampleURL, releaseTars string) {
+	var title string
+	if *preview {
+		title = "Branch "
+	}
+
+	if releaseTag == "HEAD" || releaseTag == branchHead {
+		title += branch
+	} else {
+		title += releaseTag
+	}
+
+	if *preview {
+		f.WriteString(fmt.Sprintf("**Release Note Preview - generated on %s**\n", time.Now().Format("Mon Jan  2 15:04:05 MST 2006")))
+	}
+
+	f.WriteString(fmt.Sprintf("\n# %s\n\n", title))
+	f.WriteString(fmt.Sprintf("[Documentation](%s) & [Examples](%s)\n\n", docURL, exampleURL))
+
+	if releaseTars != "" {
+		f.WriteString(fmt.Sprintf("## Downloads for %s\n\n", title))
+		createDownloadsTable(f, releaseTag, "", releaseTars+"/kubernetes.tar.gz", releaseTars+"/kubernetes-src.tar.gz")
+		createDownloadsTable(f, releaseTag, "Client Binaries", releaseTars+"/kubernetes-client*.tar.gz")
+		createDownloadsTable(f, releaseTag, "Server Binaries", releaseTars+"/kubernetes-server*.tar.gz")
+		createDownloadsTable(f, releaseTag, "Node Binaries", releaseTars+"/kubernetes-node*.tar.gz")
+		f.WriteString("\n")
+	}
+}
+
+// createDownloadTable creates table of download link and sha256 hash for given file.
+func createDownloadsTable(f *os.File, releaseTag, heading string, filename ...string) {
+	var urlPrefix string
+
+	if *releaseBucket == "kubernetes-release" {
+		urlPrefix = k8sReleaseURLPrefix
+	} else {
+		urlPrefix = fmt.Sprintf("https://storage.googleapis.com/%s/release", *releaseBucket)
+	}
+
+	if heading != "" {
+		f.WriteString(fmt.Sprintf("\n### %s\n", heading))
+	}
+
+	f.WriteString("\n")
+	f.WriteString("filename | sha256 hash\n")
+	f.WriteString("-------- | -----------\n")
+
+	files := make([]string, 0)
+	for _, name := range filename {
+		fs, _ := filepath.Glob(name)
+		for _, v := range fs {
+			files = append(files, v)
+		}
+	}
+
+	for _, file := range files {
+		fn := filepath.Base(file)
+		sha, err := u.GetSha256(file)
+		if err != nil {
+			log.Printf("failed to calc SHA256 of file %s: %s", file, err)
+			os.Exit(1)
+		}
+		f.WriteString(fmt.Sprintf("[%s](%s/%s/%s) | `%s`\n", fn, urlPrefix, releaseTag, fn, sha))
+	}
+}
+
+// minorReleases performs a minor (vX.Y.0) release by fetching the release template and aggregate
+// previous release in series.
+func minorRelease(f *os.File, release, draftURL, changelogURL string) {
+
+	// Check for draft and use it if available
+	log.Printf("Checking if draft release notes exist for %s...", release)
+
+	resp, err := http.Get(draftURL)
+	if err == nil {
+		defer resp.Body.Close()
+	}
+
+	// TODO: find a better way to tell failed response
+	if err == nil && (resp.StatusCode == 200 || resp.StatusCode == 304) {
+		log.Printf("Draft found - using for release notes...")
+		_, err = io.Copy(f, resp.Body)
+		if err != nil {
+			log.Printf("error during copy to file: %s", err)
+			return
+		}
+		f.WriteString("\n")
+	} else {
+		log.Printf("No draft found - creating generic template...")
```

Comment:
"No draft found" sounds like the error reason is "There is not such thing." What if it failed because of timeout, or authentication issue? You better say "Failed to do ..." or "Error when ..." plus printing out actual error message you get. Or you can return this error to upper level, and print it out there.

Reply:
Done